### DISCUSSION TITLE
/articles/sharepoint-framework-use-npm-packages-to-share-code-in-spfx-projects-not-library-components/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#235] [Comments]
Curious about sharing code in SharePoint Framework projects? Explore the advantages of npm packages over SPFx library components!
https://www.andrewconnell.com/articles/sharepoint-framework-use-npm-packages-to-share-code-in-spfx-projects-not-library-components/

**Comment by StfBauer** [comment]
Yeah, same here. They are pretty useless. I wrote a blog post about those almost five years ago: https://n8d.at/tips-and-tricks-working-with-spfx-library-components

Finally, I found a way to overcome many hurdles even without publishing packages to an npm registry.

**Reply by vishalshitole** [reply]
@StfBauer Curious to know the steps you followed to use the package without publishing it to an npm registry.

**Reply by andrewconnell** [reply]
@vishalshitole  You can point to a folder in your project...

```console
npm install ../other-folder-containing-package
```

... or a tarball...

```console
npm install ./package.tgz
```

... or a git repo...

```console
npm install mygithubuser/myproject
npm install github:mygithubuser/myproject
```

... or many other options [documented in for npm install](https://docs.npmjs.com/cli/v8/commands/npm-install)

**Comment by ynot3363** [comment]
Another con for SPFx Library Components is they make CI/CD pipelines ridiculous to implement. Laura did a great job documenting how you can do this: https://laurakokkarinen.com/how-to-implement-azure-devops-pipelines-for-spfx-solutions-that-use-library-components/, although I did run into an error and that was related to the fact that npm install fails because it cannot find the library-component. 

Temp solution right now is to dynamically modify the package.json file with a postinstall script to add in the library-component dependency. 

Time to look into npm packages

**Reply by andrewconnell** [reply]
![](https://gifdb.com/images/high/preach-can-i-get-an-amen-will-ferrell-ai9tnpleabvcsl6s.gif)

**Comment by heinrich-ulbricht** [comment]
Last time I tried using the npm approach was about 5 (or more) years ago. Library components didn't exist then. I remember it being hard to debug the code in VS Code. It was like: either I can step into the web part code OR I can step into the npm bundle code. But not both, making debugging a headache.

Is that a problem you faced? Because since this experience I stayed away from the npm package approach. Maybe time to revisit?

**Reply by andrewconnell** [reply]
Little unclear... not sure if you're talking about SPFx components or NPM packages with the debugging challenge. I don't have an issue with this... but then again, I mostly do my debugging for client side dev in the browser's dev tools and not VS Code as I find VS Code loses the connection frequently (not just an SPFx thing).

That's partly due to the fact I have multiple browsers open and when you start a debugging session, you usually need to kill all instances of that browser that you use for debugging (as VSCode can't spin up a new one an open the debugging port 9222 on the browser if there are already instances open).

**Reply by heinrich-ulbricht** [reply]
The npm package approach did not work well when using VS Code for debugging. Ok, I get that you basically worked around all that headache by debugging in the browser. I'm resorting to this myself when VS Code once again won't connect. Thanks for your feedback.

**Reply by andrewconnell** [reply]
You bet!
